src/disk_cache.rs

```diff
@@ -0,0 +1,1132 @@
+use std::{collections::HashMap, fmt::Display, ops::Range, sync::Arc};
+
+use bytes::{Bytes, BytesMut};
+use futures::{future::BoxFuture, stream, stream::BoxStream, StreamExt};
+use object_store::{
+    buffered::BufWriter, Attribute, Attributes, GetRange, GetResultPayload, PutResult,
+};
+use object_store::{path::Path, GetOptions, GetResult, ObjectMeta, ObjectStore};
+use rand::{distributions::Alphanumeric, Rng};
+use serde::{Deserialize, Serialize};
+use tokio::{
+    fs,
+    fs::{File, OpenOptions},
+    io::{AsyncReadExt, AsyncWriteExt},
+};
+
+use crate::metrics::DbStats;
+
+pub(crate) struct CacheableGetOptions {
+    /// the range of the object to get, if None, the entire object will be fetched.
+    pub range: Option<GetRange>,
+    /// the cache strategy to use for this request, you can choose to
+    /// disable caching.
+    pub cache_enabled: bool,
+    /// the lower the rank, the less likely the object will be evicted.
+    /// if set to Some(0), the object will never be evicted.
+    #[allow(unused)]
+    pub cache_rank: Option<u8>,
+}
+
+impl Default for CacheableGetOptions {
+    fn default() -> Self {
+        Self {
+            range: None,
+            cache_enabled: true,
+            cache_rank: None,
+        }
+    }
+}
+
+/// CachableObjectStore is a wrapper around ObjectStore, it caches the object
+/// in the local, and serves the object from the cache if possible.
+#[derive(Clone, Debug)]
+pub(crate) enum CacheableObjectStoreRef {
```

Comment:
I think you answered this elsewhere (and it was some Rust idiosyncrasy) but curious why not a `trait` and implement two different structs, a cached and a direct?